framework/src/mesh/MooseMesh.C

```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
   return neighbor_elems;
 }
 
+bool
+MooseMesh::isBoundaryInternalToSubdomains(BoundaryID bid,
+                                          const std::set<SubdomainID> & blk_group) const
+{
+  mooseAssert(_bnd_elem_range, "Boundary element range is not initialized");
+  const bool all_blocks =
+      std::find(blk_group.begin(), blk_group.end(), Moose::ANY_BLOCK_ID) != blk_group.end();
+
+  // Loop over all side elements of the mesh, select those on the boundary
+  for (const auto & bnd_elem : *_bnd_elem_range)
+  {
+    const auto & [elem_ptr, elem_side, elem_bid] = *bnd_elem;
+    if (elem_bid == bid)
+    {
+      // If an element is internal to the group of subdomain, check the neighbor
+      if (all_blocks || std::find(blk_group.begin(), blk_group.end(), elem_ptr->subdomain_id()) !=
+                            blk_group.end())
+      {
+        const auto * const neighbor = elem_ptr->neighbor_ptr(elem_side);
+
+        // If we did not ghost the neighbor, we cannot decide
+        if (neighbor == libMesh::remote_elem)
+          mooseError("Insufficient level of geometrical ghosting to determine "
+                     "if a boundary is internal to the mesh");
+        // If the neighbor does not exist, then we are on the edge of the mesh
+        if (!neighbor)
+          continue;
+        // If the neighbor is also in the group of subdomain,
+        // then the boundary cuts the subdomains
+        if (all_blocks || std::find(blk_group.begin(), blk_group.end(), neighbor->subdomain_id()) !=
+                              blk_group.end())
+          return true;
+      }
+    }
+  }
+  return false;
```

Comment:
It almost feels like this logic is the opposite of what you want. It seems like if you ever encounter a case in which you don't have a neighbor or either the element or neighbor subdomains don't match the subdomains in the provided set that you should immediately return `false`. And then only if you examine all the triples for which the boundary ID matches the provided `bid` and they meet the internal criteria would you return `true`

Reply:
I think the concept in this method is appropriate if you name this method something like `isAnyPartOfBoundaryInternalToSubdomains`

Reply:
Let's reverse it and say external